File: backend/src/services/complianceCertBlobs.ts
/**
 * Build searchable blobs for certification scoring (vendor attestation + org portal insights).
 */

export function collectComplianceUploadFileNames(payload: Record<string, unknown>): string[] {
  const names: string[] = [];
  const docUploads =
    payload.document_uploads && typeof payload.document_uploads === "object"
      ? (payload.document_uploads as Record<string, unknown>)
      : payload.documentUpload && typeof payload.documentUpload === "object"
        ? (payload.documentUpload as Record<string, unknown>)
        : null;
  if (docUploads?.["2"] && typeof docUploads["2"] === "object" && !Array.isArray(docUploads["2"])) {
    const slot2 = docUploads["2"] as Record<string, unknown>;
    const byCategory =
      slot2.byCategory && typeof slot2.byCategory === "object"
        ? (slot2.byCategory as Record<string, unknown>)
        : {};
    Object.values(byCategory).forEach((arr) => {
      if (Array.isArray(arr)) {
        arr.forEach((name) => {
          if (typeof name === "string" && name.trim()) names.push(name.trim());
        });
      }
    });
  }
  return names;
}

export function certificationFormTextFromGetter(get: (k: string) => unknown): string {
  const parts: string[] = [];
  const walk = (v: unknown) => {
    if (v == null) return;
    if (Array.isArray(v)) {
      v.forEach(walk);
      return;
    }
    if (typeof v === "object") {
      Object.values(v as Record<string, unknown>).forEach(walk);
      return;
    }
    const t = String(v).trim();
    if (t) parts.push(t);
  };
  walk(get("security_certifications"));
  walk(get("security_compliance_certificates"));
  walk(get("regulatorycompliance_cert_material"));
  return parts.join(" ");
}

export function buildCertificationsSearchBlobsFromPayload(payload: Record<string, unknown>): {
  certificationsSearchBlob: string;
  complianceUploadBlob: string;
} {
  const complianceUploadNames = collectComplianceUploadFileNames(payload);
  const complianceUploadBlob = complianceUploadNames.join(" ").toLowerCase();
  const get = (k: string) => payload[k];
  const certFormBlob = certificationFormTextFromGetter(get).toLowerCase();
  return {
    certificationsSearchBlob: `${certFormBlob} ${complianceUploadBlob}`.trim(),
    complianceUploadBlob,
  };
}

/** Flatten buyer-stated vendor certifications (array / JSON string / plain string) for scoring. */
export function buyerVendorCertificationsToSearchBlob(vendorCertifications: unknown): string {
  const parts: string[] = [];
  const walk = (v: unknown) => {
    if (v == null) return;
    if (Array.isArray(v)) {
      v.forEach(walk);
      return;
    }
    if (typeof v === "object") {
      Object.values(v as Record<string, unknown>).forEach(walk);
      return;
    }
    const t = String(v).trim();
    if (t) parts.push(t);
  };
  if (typeof vendorCertifications === "string") {
    const s = vendorCertifications.trim();
    if (s.startsWith("[") || s.startsWith("{")) {
      try {
        walk(JSON.parse(s));
      } catch {
        parts.push(s);
      }
    } else {
      parts.push(s);
    }
  } else {
    walk(vendorCertifications);
  }
  return parts.join(" ").toLowerCase();
}

File: python/services/compliance_cert_blobs.py
"""Build searchable blobs for certification scoring."""

from __future__ import annotations

from typing import Any, Callable


def collect_compliance_upload_file_names(payload: dict[str, Any]) -> list[str]:
    names: list[str] = []
    doc_uploads = None
    if isinstance(payload.get("document_uploads"), dict):
        doc_uploads = payload["document_uploads"]
    elif isinstance(payload.get("documentUpload"), dict):
        doc_uploads = payload["documentUpload"]
    if not doc_uploads:
        return names
    slot2 = doc_uploads.get("2")
    if not isinstance(slot2, dict):
        return names
    by_category = slot2.get("byCategory") if isinstance(slot2.get("byCategory"), dict) else {}
    for arr in by_category.values():
        if isinstance(arr, list):
            for name in arr:
                if isinstance(name, str) and name.strip():
                    names.append(name.strip())
    return names


def certification_form_text_from_getter(get: Callable[[str], Any]) -> str:
    parts: list[str] = []

    def walk(v: Any) -> None:
        if v is None:
            return
        if isinstance(v, list):
            for item in v:
                walk(item)
            return
        if isinstance(v, dict):
            for item in v.values():
                walk(item)
            return
        t = str(v).strip()
        if t:
            parts.append(t)

    walk(get("security_certifications"))
    walk(get("security_compliance_certificates"))
    walk(get("regulatorycompliance_cert_material"))
    hipaa_baa = str(get("hipaa_baa") or "").strip().lower()
    if hipaa_baa in ("yes", "yes_standard", "yes_on_request"):
        parts.append("HIPAA BAA")
    return " ".join(parts)